File: repos/app/wrapper/lcontest/LongContestUserWrapper.py
# Copyright (C) - 2014 TopCoder Inc., All Rights Reserved.

"""
Wrapper for Python MM solution.

<p>
Changes in version 1.1 (TopCoder Competition Engine - Stack Size Configuration For MM Problems v1.0):
<ol>
    <li>Added sys.setrecursionlimit(10**7) in order to avoid recursion limit problems in user code.</li>
</ol>
</p>

@author Selena
@version 1.1
"""

import re
import struct 
import os
import sys
import time
import <EXPOSED_WRAPPER_CLASS>

sys.setrecursionlimit(10**7)

class myStdout:
	def __init__(self):
		self.readAll = False
		
	def close(self):
		sys.__stdout__.close()
	def flush(self):
		sys.__stdout__.flush()
	def fileno(self):
		return sys.__stdout__.fileno()
	def isatty(self):
		return sys.__stdout__.isatty()
	def next(self):
		return sys.__stdout__.next()
	def read(self,size=None):
		return sys.__stdout__.read(size)
	def readline(self,size=None):
		return sys.__stdout__.readline(size)
	def readlines(self,size=None):
		return sys.__stdout__.readlines(size)
	def xreadlines(self):
		return sys.__stdout__.xreadlines()
	def seek(self, offset, whence=None):
		sys.__stdout__.seek(offset, whence)
	def tell(self):
		return sys.__stdout__.tell()
	def truncate(self, size=None):
		return sys.__stdout__.truncate(size)
	def write(self,str):
		if(self.readAll):
			return
		try:
			sys.__stdout__.write(str)
		except IOError:
			self.readAll = True
	def writelines(self,str):
		if(self.readAll):
			return
		try:
			sys.__stdout__.writelines(str)
		except IOError:
			self.readAll = True
	
#sys.stdout = myStdout()

SIG_PROF = 27
INT_TYPE = 1
LONG_TYPE = 2
DOUBLE_TYPE = 3
STRING_TYPE = 4
VEC_INT_TYPE = 5
VEC_LONG_TYPE = 6
VEC_DOUBLE_TYPE = 7
VEC_STRING_TYPE = 8

EXCEPTION = 250
TIMEOUT = 251
ABORT = 252
METHOD_START = 253
TIME = 254
TERMINATE = 255

#library code

def flush(fd):
	try: os.fsync(fd)
	except OSError, e:
		pass

def writeInt2(fd, a):
	os.write(fd,chr((a>>24)&255))
	os.write(fd,chr((a>>16)&255))
	os.write(fd,chr((a>>8)&255))
	os.write(fd,chr(a&255))

def writeInt(fd, a):
	os.write(fd,chr(INT_TYPE))
	writeInt2(fd,a)

def startMethod(fd, a):
	os.write(fd,chr(METHOD_START))
	writeInt2(fd,a)

def writeTime(fd, a):
	os.write(fd,chr(TIME))
	writeInt2(fd,a)

def writeLong(fd, a):
	os.write(fd,chr((a>>56)&255))
	os.write(fd,chr((a>>48)&255))
	os.write(fd,chr((a>>40)&255))
	os.write(fd,chr((a>>32)&255))
	os.write(fd,chr((a>>24)&255))
	os.write(fd,chr((a>>16)&255))
	os.write(fd,chr((a>>8)&255))
	os.write(fd,chr(a&255))

def writeLong(fd, a):
	os.write(fd,chr(LONG_TYPE))
	writeLong2(fd,a)

def writeDouble2(fd, a):
	os.write(fd,struct.pack(">d",a))

def writeDouble(fd, a):
	os.write(fd,chr(DOUBLE_TYPE))
	writeDouble2(fd,a)

def writeString2(fd, a):
	writeInt2(fd,len(a))
	os.write(fd,a)

def writeString(fd, a):
	os.write(fd,chr(STRING_TYPE))
	writeString2(fd,a)
	
def writeIntArray(fd, a):
	os.write(fd,chr(VEC_INT_TYPE))
	writeInt2(fd,len(a))
	for i in a :
		writeInt2(fd,i)

def writeLongArray(fd, a):
	os.write(fd,chr(VEC_LONG_TYPE))
	writeInt2(fd,len(a))
	for i in a :
		writeLong2(fd,i)

def writeDoubleArray(fd, a):
	os.write(fd,chr(VEC_DOUBLE_TYPE))
	writeInt2(fd,len(a))
	for i in a :
		writeDouble2(fd,i)

def writeStringArray(fd, a):
	os.write(fd,chr(VEC_STRING_TYPE))
	writeInt2(fd,len(a))
	for i in a :
		writeString2(fd,i)
	
def writeArg(fd, a):
	if(isinstance(a,list)):
		if(len(r) == 0):
			os.write(fd,VEC_INT_TYPE)
			writeInt2(fd,0)
		elif(isinstance(a[0],int)):
			writeIntArray(fd, a)
		elif(isinstance(a[0],long)):
			writeLongArray(fd, a)
		elif(isinstance(a[0],double)):
			writeDoubleArray(fd, a)
		elif(isinstance(a[0],str)):
			writeStringArray(fd, a)
	elif(isinstance(a,int)):
		writeInt(fd, a)
	elif(isinstance(a,long)):
		writeLong(fd, a)
	elif(isinstance(a,double)):
		writeDouble(fd, a)
	elif(isinstance(a,str)):
		writeString(fd, a)

def getInt2(fd):
	ret = struct.unpack(">i",os.read(fd,4))
	return ret[0]

def getInt(fd):
	os.read(fd,1)
	return getInt2(fd)

def getLong2(fd):
	return struct.unpack(">q",os.read(fd,8))[0]

def getLong(fd):
	os.read(fd,1)
	return getLong2(fd)

def getDouble2(fd):
	return struct.unpack(">d",os.read(fd,8))[0]

def getDouble(fd):
	os.read(fd,1)
	return getDouble2(fd)

def getString2(fd):
	L = getInt2(fd)
	res = ''
	while L > 0:
		s = os.read(fd,L)
		res = res + s
		L = L - len(s)

	return res

def getString(fd):
	os.read(fd,1)
	return getString2(fd)

def getIntArray(fd):
	os.read(fd,1)
	len = getInt2(fd)
	ret = []
	i = 0
	while( i < len ):
		ret.append(getInt2(fd))
		i = i + 1
	return ret
		
def getLongArray(fd):
	os.read(fd,1)
	len = getInt2(fd)
	ret = []
	i = 0
	while( i < len ):
		ret.append(getLong2(fd))
		i = i + 1
	return ret
		
		
def getDoubleArray(fd):
	os.read(fd,1)
	len = getInt2(fd)
	ret = []
	i = 0
	while( i < len ):
		ret.append(getDouble2(fd))
		i = i + 1
	return ret
		
		
def getStringArray(fd):
	os.read(fd,1)
	len = getInt2(fd)
	ret = []
	i = 0
	while( i < len ):
		ret.append(getString2(fd))
		i = i + 1
	return ret
		
		

getchBuffer = ''
getchPos = -1
getchBufferSize = 0
totalTm = 0

def getch(fd):
	global getchPos
	global getchBuffer
	
	if(getchPos >= len(getchBuffer)):
		getchPos = -1
		
	if(getchPos == -1):
		getchBuffer = os.read(fd,1000000)
		getchPos = 0
		
	ret = getchBuffer[getchPos]
	getchPos = getchPos + 1
	
	return ret

def gettime():
	r = time.time()
	r = r * 1000
	
	return long(r)
		
#wrapper starts here

def gettimefloat():
	r = time.time()
	r = r * 1000
	
	return r

def main():
	#global totalTm
	
	initialized = False
	
	fd_out = 12
	fd_in = 11
	
	mod = ""
	sol = ""

	timercnt = 0
	
	mypid = os.getpid()
	while 1:
		
		command = ord(os.read(fd_in,1))
		#print command
		
		if(command == TERMINATE or command == -1):
			sys.exit(0)
		elif(command == TIMEOUT):
			getInt2(fd_in)
		elif(command == METHOD_START):
			method = getInt2(fd_in)
			if(method == -1):
				pass
			<METHODS>
			elif(method == <METHOD_NUMBER>):
				#tm2 = gettime()
				
				<ARGS>
				<ARG_NAME> = <ARG_METHOD_NAME>(fd_in)
				</ARGS>
								
				#print totalTm
				
				os.kill(mypid, SIG_PROF)
				
				tm = gettime()
				timercnt = timercnt + 1
				
				if(not initialized):
					mod = __import__('<CLASS_NAME>')
					sol = mod.<CLASS_NAME>()
					initialized = True
	
				ret = sol.<METHOD_NAME>(<PARAMS>)
								
				tm = gettime() - tm - <EXPOSED_WRAPPER_CLASS>.watch.getTime()
				timercnt = timercnt + 1
						
				if(tm > 0 and timercnt >= 50):
					tm = tm - 1
					timercnt = timercnt - 50
		
				os.kill(mypid, SIG_PROF)
				
				writeTime(fd_out, tm)
				writeArg(fd_out, 1)
				<RET_METHOD_NAME>(fd_out, ret)
				flush(fd_out)
				
				sys.__stdout__.flush()
				sys.__stderr__.flush()
				
				
			</METHODS>


if __name__ == "__main__":
    main()

File: repos/app/wrapper/lcontest/LongContestUserWrapper.R
library(compiler)
library(methods)
library(R.oo)
library(R.utils)
library(fork)
dyn.load("<C_HELPER>")

INT_TYPE        <- as.integer(1)
LONG_TYPE       <- as.integer(2)
DOUBLE_TYPE     <- as.integer(3)
STRING_TYPE     <- as.integer(4)
VEC_INT_TYPE    <- as.integer(5)
VEC_LONG_TYPE   <- as.integer(6)
VEC_DOUBLE_TYPE <- as.integer(7)
VEC_STRING_TYPE <- as.integer(8)

EXCEPTION       <- as.integer(250)
TIMEOUT         <- as.integer(251)
ABORT           <- as.integer(252)
METHOD_START    <- as.integer(253)
TIME            <- as.integer(254)
TERMINATE       <- as.integer(255)

fin <- as.integer(11)
fout <- as.integer(12)

flush <- function() {
    .C("flush",fout)
}

flush2 <- function(fd) {
    .C("flush",fd)
}

writeSingleInt <- function(i) {
    .C("writeSingleInt",fout,i)
}

writeSingleInt2 <- function(fd,i) {
    .C("writeSingleInt",fd,i)
}

writeInt2 <- function(i) {
    .C("writeInt",fout,i)
}

writeInt3 <- function(fd,i) {
    .C("writeInt",fd,i)
}

writeInt <- function(i) {
    .C("writeSingleInt",fout,INT_TYPE)
    writeInt2(i)
}

writeLong <- function(ll) {
    .C("writeSingleInt",fout,LONG_TYPE)
    .C("writeLongLong",fout,ll)
}

writeDouble <- function(d) {
    .C("writeSingleInt",fout,DOUBLE_TYPE)
    .C("writeDouble",fout,d)
}

writeString2 <- function(str) {
    strLen <- nchar(str)
    writeInt2(strLen)
    for(n in 1:strLen) {
        everyChar <- charToInt(substr(str,n,n))
        writeSingleInt(as.integer(everyChar))
    }
}

writeString <- function(str) {
    .C("writeSingleInt",fout,STRING_TYPE)
    writeString2(str)
}

writeIntArray <- function(i_arr) {
    .C("writeSingleInt",fout,VEC_INT_TYPE)
    len <- length(i_arr)
    for(n in 1:len) {
        writeInt(i_arr[n])
    }
}

writeLongArray <- function(l_arr) {
    .C("writeSingleInt",fout,VEC_LONG_TYPE)
    len <- length(l_arr)
    for(n in 1:len) {
        writeLong(l_arr[n])
    }
}

writeDoubleArray <- function(d_arr) {
    .C("writeSingleInt",fout,VEC_DOUBLE_TYPE)
    len <- length(d_arr)
    for(n in 1:len) {
        writeLong(d_arr[n])
    }
}

writeStringArray <- function(str_arr) {
    .C("writeSingleInt",fout,VEC_STRING_TYPE)
    len <- length(str_arr)
    writeInt2(len)
    for(n in 1:len) {
        writeString2(str_arr[n])
    }
}

writeMethod <- function(i) {
    writeSingleInt(METHOD_START)
    writeInt2(i)
}

writeTime <- function(t) {
    writeSingleInt(TIME)
    writeInt2(t)
}

writeException <- function(e) {
    writeSingleInt(EXCEPTION)
    writeInt2(e)
}

writeArg <- function(val,type) {
    if(type == INT_TYPE) {
        writeInt(val)
    } else if(type == LONG_TYPE) {
        writeLong(val)
    } else if(type == DOUBLE_TYPE) {
        writeDouble(val)
    } else if(type == VEC_INT_TYPE) {
        writeIntArray(val)
    } else if(type == VEC_LONG_TYPE) {
        writeLongArray(val)
    } else if(type == VEC_DOUBLE_TYPE) {
        writeDoubleArray(val)
    } else if(type == VEC_STRING_TYPE) {
        writeStringArray(val)
    }
}

getInt <- function() {
    s1 <- .C("getSingleInt",fin,result=as.integer(10))
    return (s1$result)
}

getInt2 <- function() {
    s2 <- .C("getInt2",fin,result=as.integer(12))
    return (s2$result)
}

getLong2 <- function() {
    s2 <- .C("getLongLong2",fin,result=as.numeric(12))
    return (s2$result)
}

getLong <- function() {
    type <- getInt()
    return (getLong2())
}

getDouble2 <- function() {
    s2 <- .C("getDouble2",fin,result=as.double(12))
    return (s2$result)
}

getDouble <- function() {
    type <- getInt()
    return (getDouble2())
}

getString2 <- function() {
    b <- as.character("")
    len <- getInt2()
    for(n in 1:len) {
            c <- getInt()
            b <- paste(b,intToChar(c),sep="")
    }
    return (b)
}
getString <- function() {
    type <- getInt()
    return (getString2())
}

getStringArray <- function() {
    ret <- list()
    type <- getInt()
    strLen <- getInt2()
    for(n in 1:strLen) {
        line <- getString2()
        ret[n] <- line
    }
    return (ret)
}

getIntArray <- function() {
    type <- getInt()
    i_len <- getInt2()
    ret <- list()
    for(n in 1:i_len) {
        ret[n] <- getInt2()
    }
    return (ret)
}

getDoubleArray <- function() {
    type <- getInt()
    d_len <- getInt2()
    ret <- list()
    for(n in 1:d_len) {
        ret[n] <- getDouble2()
    }
    return (ret)
}

getLongArray <- function() {
    type <- getInt()
    l_len <- getInt2()
    ret <- list()
    for(n in 1:l_len) {
        ret[n] <- getLong2()
    }
    return (ret)
}

gettime <- function() {
    return (System$currentTimeMillis())
}

startMethod <- function(fd, a) {
    writeSingleInt2(fd,METHOD_START)
    writeInt3(fd,a)
}
main <- function() {
    initialized <- FALSE
    mod <- ""
    sol <- ""
    timercnt <- as.integer(0)
    pid <- getpid()
    while ( TRUE ) {
        command <- getInt()
        if(command == TERMINATE || command == -1) {
            exit(0)
        } else if(command == TIMEOUT) {
            d <- getInt2()
        } else if(command == METHOD_START) {
            method <- getInt2()
            if(method == -1) {
                #do nothing
            } 
            <METHODS>
            else if(method == <METHOD_NUMBER>) {
                <ARGS>
                    <ARG_NAME> = <ARG_METHOD_NAME>()
                </ARGS>
				
				
                kill(pid, signal = 27)
                tm <- gettime()

                timercnt <- timercnt + 1
                if(!initialized) {
                    solPath <- paste(scriptPath,"/<CLASS_NAME>.rlc",sep="")
                	loadcmp(solPath)
                	sol <- new ("<CLASS_NAME>", <PARAMS>)
                    initialized = TRUE
                }
                
                sol <- initialize(sol,<PARAMS>)

                ret <- <METHOD_NAME>(sol)

				solPath <- paste(scriptPath,"/<EXPOSED_WRAPPER_CLASS>.rlc",sep="")
                loadcmp(solPath)

				tm <- gettime() - tm - getExposedTime()
				
                timercnt <- timercnt + 1

                if(tm > 0 && timercnt >= 50) {
                    tm <- tm - 1
                    timercnt <- timercnt - 50
                }

                kill(pid, signal = 27)

                writeTime(as.integer(tm))
                writeArg(as.integer(1),INT_TYPE)
                <RET_METHOD_NAME>(ret)
                flush()
            }
            </METHODS>
        }
    }
}
main()